corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -305,23 +307,33 @@ def _get_form_ids_for_user(domain, user_id, is_deleted):
 
     @staticmethod
     def get_all_forms_received_since(received_on_since=None, chunk_size=500):
-        return _batch_iterate(
-            batch_fn=FormAccessorSQL.get_forms_received_since,
+        """Iterate through all forms in the entire database, optionally received since
+        a specific date
+        """
+        return _batch_iterate_over_shards(
+            batch_fn=FormAccessorSQL._get_forms_received_since,
             next_start_from_fn=lambda form: form.received_on,
             start_from=received_on_since,
             chunk_size=chunk_size
         )
 
     @staticmethod
-    def get_forms_received_since(received_on_since=None, limit=500):
+    def _get_forms_received_since(received_on_since=None, limit=500, using=None):
+        """
+        Function to be used with ``_batch_iterate_over_shards`` to iterate over all
+        forms in the entire database.
+
+        :param using: alternate DB alias to query against
+        :return:
+        """
         received_on_since = received_on_since or datetime.min
         results = XFormInstanceSQL.objects.raw(
             'SELECT * FROM get_all_forms_received_since(%s, %s)',
-            [received_on_since, limit]
+            [received_on_since, limit],
+            using=using
```

Comment:
This is the main change in each function. By telling the query which db alias to use we can query the shard DBs individually and hence don't end up getting N x the limit (N = num shards) on each query.